libs/user/user.ts

```diff
@@ -1,93 +1,160 @@
-    const ethAddress = getEthAddressFromPrivateKey(
-      config.ethereumPrivateKey,
-      this.web3Websocket.web3,
-    );
-    if (!ethAddress) return null;
-    this.ethPrivateKey = config.ethereumPrivateKey;
-    this.ethAddress = ethAddress;
-
-    // Set this.nightfallMnemonic, this.zkpKeys,
-    // subscribe to incoming viewing keys if valid mnemonic,
-    // or creates one if no mnemonic was given
-    const nightfallKeys = await createZkpKeysFromMnemonic(
-      config.nightfallMnemonic,
-      this.client,
+    if (!shieldContractAddress)
+      throw new Error("Unable to get Shield contract address");
+
+    // Get the Eth account address from private key if it's a valid key
+    const ethAddress = getEthAccountAddress(ethPrivateKey, web3Websocket.web3);
+    if (!ethAddress) throw new Error("Unable to get an Eth address");
+
+    // Create a set of Zero-knowledge proof keys from a valid mnemonic
+    // or from a new mnemonic if none was provided,
+    // subscribe to incoming viewing keys
+    const nightfallKeys = await createZkpKeysAndSubscribeToIncomingKeys(
+      nightfallMnemonic,
+      client,
     );
-    if (!nightfallKeys) return null;
-    const { nightfallMnemonic, zkpKeys } = nightfallKeys;
-    this.nightfallMnemonic = nightfallMnemonic;
-    this.zkpKeys = zkpKeys;
+    if (!nightfallKeys) throw new Error("Unable to generate Nightfall keys");
+
+    return new User({
+      client,
+      web3Websocket,
+      shieldContractAddress,
+      ethPrivateKey: ethPrivateKey,
+      ethAddress,
+      nightfallMnemonic: nightfallKeys.nightfallMnemonic,
+      zkpKeys: nightfallKeys.zkpKeys,
+    });
+  }
+}
 
-    return { User: this };
+class User {
+  // Set by constructor
+  client: Client;
+  web3Websocket: Web3Websocket;
+  shieldContractAddress: string;
+  ethPrivateKey: string;
+  ethAddress: string;
+  nightfallMnemonic: string;
+  zkpKeys: NightfallZkpKeys;
+
+  // Set when transacting
+  token: any;
+  nightfallTxHashes: string[] = [];
+
+  constructor(options: UserOptions) {
+    logger.debug("new User");
+
+    let key: keyof UserOptions;
+    for (key in options) {
+      this[key] = options[key];
+    }
   }
 
-  // TODO needs massive refactor
-  async makeDeposit(options: UserDeposit): Promise<any> {
+  async makeDeposit(options: UserMakeDepositOptions) {
     logger.debug({ options }, "User :: makeDeposit");
-
-    return createDeposit(
-      options.tokenAddress,
-      options.tokenStandard,
-      options.value,
-      options.fee || TX_FEE_DEFAULT,
-      this.shieldContractAddress,
+    makeDepositOptions.validate(options);
+
+    // Format options
+    let value = options.value.trim();
+    let fee = options.feeGwei?.trim() || TX_FEE_GWEI_DEFAULT;
+    const tokenAddress = options.tokenAddress.trim();
+    const tokenStandard = options.tokenStandard.trim().toUpperCase();
+
+    // Set token only if it's not set or is different
+    if (!this.token || tokenAddress !== this.token.contractAddress)
```

Comment:
I would surround the instruction with `{}`, this is a `big` instruction that spans multiple lines, just to avoid confusion.

Reply:
True :memo: